### PR TITLE
operator: Add OpenShift CloudCredentials support for AWS STS

### PR DESCRIPTION
**Pull request by JoaoBraveCoding**
**What this PR does / why we need it**:
The following PR adds auxiliary automation to reconcile the cloud credentials secret for OpenShift clusters installations on AWS using STS as the default method for cloud credentials. In detail the installation of the Loki Operator requires from the cluster administrator to provide a AWS Role ARN that will be used by the operands (i.e. Loki) to request short-lived token from STS to access S3. The enablement is based on the following OpenShift proposals:
- [Tokenized Authentication for Red Hat Operators on Cloud Providers](https://github.com/openshift/enhancements/blob/master/enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md)
- openshift/enhancements#1503

Effectively the underlying implementation utilizes the `CredentialsRequest` using the Loki Operator AWS Role ARN to request the credentials secret from the OpenShift cloud-credentials-operator. It minimizes user input on object storage access  for LokiStack installation to a list of buckets and an AWS region.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:
Depends on:
- grafana/loki#11481
- grafana/loki#11592

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] Documentation added
- [x] Tests updated
- [x] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: operator/go.mod [hunk 1]
```diff
@@ -11,23 +11,25 @@ require (
...
 	github.com/stretchr/testify v1.8.4
 	gopkg.in/yaml.v2 v2.4.0
-	k8s.io/api v0.27.7
-	k8s.io/apimachinery v0.27.7
-	k8s.io/apiserver v0.27.7
-	k8s.io/client-go v0.27.7
+	k8s.io/api v0.28.3
+	k8s.io/apimachinery v0.28.3
+	k8s.io/apiserver v0.28.3
+	k8s.io/client-go v0.28.3
```
Comment (periklis) [inline operator/go.mod:25]: Do we need to upgrade? Or can we use `github.com/openshift/cloud-credential-operator@release-4.14` for now?

File: operator/internal/manifests/openshift/credentialsrequest.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	if err != nil {
+		return nil, kverrors.Wrap(err, "failed encoding credentialsrequest provider spec")
+	}
+
+	return &cloudcredentialv1.CredentialsRequest{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-%s", stack.Namespace, secretName),
+			Namespace: ccoNamespace,
+			Annotations: map[string]string{
+				AnnotationCredentialsRequestOwner: stack.String(),
```
Comment (JoaoBraveCoding) [inline operator/internal/manifests/openshift/credentialsrequest.go:35]: Question why do we use label and not something like the Owner reference https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/controller/controllerutil/controllerutil.go?

File: operator/internal/handlers/internal/storage/secrets.go [hunk 3]
```diff
@@ -19,47 +20,92 @@ import (
...
 	}
 
-	return &storageSecret, nil
+	return &storageSecret, nil, nil
 }
 
-// extractSecret reads a k8s secret into a manifest object storage struct if valid.
-func extractSecret(s *corev1.Secret, secretType lokiv1.ObjectStorageSecretType) (storage.Options, error) {
-	hash, err := hashSecretData(s)
+// extractSecrets reads a k8s secret into a manifest object storage struct if valid.
```
Comment (JoaoBraveCoding) [inline operator/internal/handlers/internal/storage/secrets.go:69]: maybe my suggestion is not perfect but something of the sorts
```suggestion
// extractSecrets reads the k8s obj storage secret into a manifest object storage struct if valid.
// The managed auth is also read into the manifest object under the right circumstances.
```

File: operator/controllers/loki/credentialsrequests_controller.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+// CredentialsRequestsReconciler reconciles a CredentialsRequest resource a LokiStack request.
+type CredentialsRequestsReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+	Log    logr.Logger
+}
+
+// Reconcile creates a new CredentialsRequest for the OpenShift cloud-credentials-operator to process
+// for a Lokistack not annotated with `loki.grafana.com/credentials-request-secret-ref`. If the LokiStack
+// resource is not found any accompanying CredentialsRequest resource is deleted if found.
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/controllers/loki/credentialsrequests_controller.go]: Thanks for pointing this out, I have refine the method comment block a bit. Should be fine now.

File: operator/internal/manifests/openshift/credentialsrequest.go (hunk 2 above)
Comment (periklis) [reply operator/internal/manifests/openshift/credentialsrequest.go:35]: As explained elsewhere we cannot use Kuberentes OwnerReferences for resources across namespaces or across scope (cluster-scope vs. namespaced).

File: operator/internal/manifests/openshift/credentialsrequest.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+						"s3:GetObject",
+						"s3:DeleteObject",
+					},
+					Effect:   "Allow",
+					Resource: "arn:aws:s3:*:*:*",
+				},
+			},
+			STSIAMRoleARN: env.AWS.RoleARN,
+		}
+		secretName = fmt.Sprintf("%s-aws-creds", stackKey)
```
[the remark this answers is not in the document]
Comment (periklis) [reply operator/internal/manifests/openshift/credentialsrequest.go:75]: :thinking:

**Review by JoaoBraveCoding — COMMENTED** [review]
lgtm (can't approve my own PR 😅)

**Review by periklis — APPROVED** [review]

**Pull request merged**